### PR TITLE
RSA Implementation using OpenSSL

### PR DESCRIPTION
**Pull request by ashbm5**
Added modifications to RSACore and relevant RSA classes to leverage OpenSSL by default, there is an option jdk.nativeRSA=false to disable. The existing option to manipulate native crypto are also supported (jdk.NativeCrypto).

### CODE REVIEW

**Comment by pshipton** [comment]
@ashbm5 please fix the copyright failures and you need a Signed-off-by in the commit message
`Signed-off-by: Alon Shalev Housfater <alonsh@ca.ibm.com>`

**Comment by pshipton** [comment]
```
12:44:57 The following files were modified and have incorrect copyrights
[Pipeline] echo
12:44:57 closed/adds/jdk/src/share/classes/jdk/crypto/jniprovider/NativeCrypto.java
12:44:57 	IBM Copyright date appears to be incorrect
[Pipeline] echo
12:44:57 closed/adds/jdk/src/share/native/jdk/crypto/jniprovider/NativeCrypto.c
12:44:57 	IBM Copyright date appears to be incorrect
[Pipeline] echo
12:44:57 jdk/src/share/classes/sun/security/rsa/RSACore.java
12:44:57 	IBM Copyright (basic) is missing from the file
[Pipeline] echo
12:44:57 jdk/src/share/classes/sun/security/rsa/RSAPrivateCrtKeyImpl.java
12:44:57 	IBM Copyright (basic) is missing from the file
[Pipeline] echo
12:44:57 jdk/src/share/classes/sun/security/rsa/RSAPublicKeyImpl.java
12:44:57 	IBM Copyright (basic) is missing from the file
```

File: closed/adds/jdk/src/share/native/jdk/crypto/jniprovider/NativeCrypto.c [hunk 1]
```diff
@@ -610,3 +610,145 @@ JNIEXPORT jint JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_GCMDecrypt
...
+BIGNUM* convertJavaBItoBN(unsigned char* in, int len);
+
+/* Create an RSA Public Key
+ * Class:     jdk_crypto_jniprovider_NativeCrypto
+ * Method:    createRSAPrivateCrtKey
+ * Signature: ([BI[BI[BI[BI[BI[BI[BI[BI)J
+ */
+JNIEXPORT jlong JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_createRSAPublicKey
+  (JNIEnv *env, jclass obj, jbyteArray n, jint nLen, jbyteArray e, jint eLen) {
+    unsigned char* nNative    = (unsigned char*)((*env)->GetPrimitiveArrayCritical(env, n, 0));
```
Comment (pshipton) [inline closed/adds/jdk/src/share/native/jdk/crypto/jniprovider/NativeCrypto.c]: All of the GetPrimitiveArrayCritical calls need error checking. See elsewhere in this file.

**Comment by pshipton** [comment]
Also note https://github.com/ibmruntimes/openj9-openjdk-jdk8/pull/227, these changes will need to be done in the same way.

**Comment by ashbm5** [comment]
Due to the OpenSSL changes in #227 we also need to support OpenSSL 1.0.2 in RSA code.  The code is nearly complete but testing is ongoing.

OpenSSL 1.0.2 and OpenSSL 1.1.1 don't have the same RSA API so the code is slightly different. We found two failures in OpenSSL 1.0.2 and are currently fixing them.

**Comment by pshipton** [comment]
@ashbm5 I think we should go ahead with this in advance of adding 1.0.2 support, and add in the 1.0.2 support afterwards.

**Pull request closed without merging**